core/src/main/java/com/linecorp/armeria/client/HttpClientFactory.java

```diff
@@ -300,25 +300,36 @@ public Object newClient(ClientBuilderParams params) {
             return delegate;
         }
 
-        if (clientType == WebClient.class) {
+        // XXX(ikhoon): Consider a common interface for HTTP clients?
+        if (clientType == WebClient.class || clientType == BlockingWebClient.class ||
+            clientType == RestClient.class) {
             final RedirectConfig redirectConfig = options.redirectConfig();
             final HttpClient delegate0;
             if (redirectConfig == RedirectConfig.disabled()) {
                 delegate0 = delegate;
             } else {
                 delegate0 = RedirectingClient.newDecorator(params, redirectConfig).apply(delegate);
             }
-            return new DefaultWebClient(params, delegate0, meterRegistry);
+            final DefaultWebClient webClient = new DefaultWebClient(params, delegate0, meterRegistry);
+            if (clientType == WebClient.class) {
+                return webClient;
+            } else if (clientType == BlockingWebClient.class) {
+                return webClient.blocking();
+            } else {
+                return webClient.asRestClient();
```

Comment:
Not a big deal since I think this was also broken beforehand anyways, but I think `ScalaRestClient` may still throw an exception (doesn't have to be handled in this PR)

```
    val restClient = server.webClient().asScalaRestClient()
    val defaultRestClient = Clients.newDerivedClient(restClient)
```

Reply:
Oops... I forgot the existence of `ScalaRestClient`. 😅

Reply:
`ScalaRestClient` can't be accessed from the core module.
We may need a new API for clients in non-core. Let me add a `DecoratingClientFactory` for it.

Reply:
Added.